DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,617 @@
+    private @system T[size] stack = void;
+    private @system int _top = 0;
+
+    T pop() @trusted {if (empty) assert(0); return stack[--_top];}
+    void push(T x) @trusted {if (full) assert(0); stack[_top++] = x;}
+    bool empty() const {return _top == 0;}
+    bool full() const {return _top == size;}
+}
+
+auto getStack(T)() @trusted {
+    Stack!T result = void;
+    result._top = 0;
+    return result;
+}
+
+import std;
+
+void main() @safe {
+    auto stack = getStack!string();
+    stack._top += 1; // not allowed after this DIP
+    writeln(stack.pop); // memory corruption: garbage string is printed
+}
+```
+
+### A custom allocator
+Here we make a simple 'bump the pointer' allocator.
+```D
+@system private ubyte[4096] heap;
+@system private int heapIndex;
+
+/// allocate an array of T
+/// never frees the memory
+T[] customAlloc(T)(int length) {
+    // round up heapIndex to multiple of T.alignof
+    heapIndex = (heapIndex + T.alignof-1) & ~(T.alignof-1);
+    auto result = cast(T[]) heap[heapIndex..heapIndex + T.sizeof*length];
+    heapIndex += heapIndex + T.sizeof*length;
+    return result;
+}
+
+import std;
+
+void main() @safe {
+    string[] strArr = customAlloc!string(3);
+    heapIndex = 0; // mess up allocator integrity! not allowed after this DIP.
+    int[] intArr = customAlloc!int(3);
+    intArr[2] = 0x8035FDF0; // overwrites string array
+    writeln(strArr[0]); // memory corruption: constructed pointer
+}
+```
+
+### A custom slice type
+Many arrays are small in length and in certain situations you do not want 4 or 8 bytes to store the length when 2 suffices.
+A safe small slice type is created below. 
+It has room for extra fields but still fits in two registers.
+```D
+struct SmallSlice(T) {
+    @system private T* ptr = null;
+    @system private ushort _length = 0;
+    ubyte[size_t.sizeof-2] extraSpace; // 2 on 32-bit, 6 on 64-bit
+
+    this(T[] slice) @trusted {
+        ptr = &slice[0];
+        assert(slice.length <= ushort.max);
+        _length = cast(ushort) slice.length;
+    }
+
+    T opIndex(size_t i) @trusted {
+        return ptr[0.._length][i];
+    }
+}
+
+import std;
+
+void main() @safe {
+    int[4] arr = [10, 20, 30, 40];
+    auto slice = SmallSlice!int(arr[]);
+    __traits(getMember, slice, "_length") = 100; // change private member
+    writeln(slice[9]); // out of bounds memory access
+}
+```
+
+### A zero-terminated string
+String literals in D are zero-terminated, but as soon as you assign it to a `const(char)*` that type information is lost and using it becomes unsafe.
+With `@system` fields we can make a `@safe` zero terminated string type.
+```D
+struct Cstring {
+    @system const(char)* ptr = null;
+
+    size_t length() const @trusted {
+        import core.stdc.string: strlen;
+        return ptr ? strlen(ptr) : 0;
+    }
+}
+
+Cstring cStringLiteral(string s)() @trusted {
+    return Cstring(s.ptr); // string literals are guaranteed zero-terminated
```

Comment:
This can be ``@safe``

```d
Cstring cStringLiteral(string s)() @safe {
    return Cstring(&s[0]);
}
```

Reply:
It can't, since it's modifying a `@system` variable. Also your version gives a range violation error on empty strings. You do remind me to mention default constructors though.

Reply:
Can `cStringLiteral` really be `@trusted`, though? You're assuming that all compile-time strings (not just literals) are zero-terminated. As far as I can tell, that's not guaranteed.

```d
import std.stdio: writeln;
import core.stdc.string: strlen;
void main()
{
    enum string e = ['f', 'o', 'o'];
    writeln(strlen(e.ptr)); /* Prints "6" for me. */
}
```

Reply:
Good catch. I think this fixes it:
```D
Cstring cStringLiteral(string s)() @trusted {
    static immutable string tmp = s;
    return Cstring(tmp.ptr); // static strings are guaranteed zero-terminated
}
```

Reply:
I think I understand where we differ.

A struct's generated constructor should be considered ``@trusted`` as far as this DIP is concerned (right now it would be ``@safe``).

You are taking a known state, setting it to another known state that was intended by the author.

Reply:
I added this:
> A struct with `@system` fields can not be constructed using the default constructor in `@safe` code.
A `@trusted` constructor can be defined to allow safe construction.

Do you think default constructors should be `@trusted` by default?

Reply:
Yes I do.
The lack of a constructor is just as much a design choice as to adding one.

If it is not the default, there would need to be a good argument against it.

For example in the above c string example, if the expression used to initialize the pointer is not safe, then the type initialization will fail. But if it is safe i.e. ``&s[0]``, then it will succeed.

Reply:
>  The lack of a constructor is just as much a design choice as to adding one.
> If it is not the default, there would need to be a good argument against it.

The point of `@trusted` code is that it explicitly marks sections that need inspection, while a hidden `@trusted` constructor makes it harder to identify places where memory corruption might happen. Also the point of `@system` fields it to place additional invariants on the fields, while a `@trusted` default constructor implies there are no extra invariants, so why would you give it `@system` fields then?

> But if it is safe i.e. &s[0], then it will succeed.

But initializing `Cstring` with `&s[0]` definitely shouldn't succeed on arbitrary `char[]`.
The point of `Cstring` is that its `@system` field may not be an arbitrary valid `const(char)*`, but only one that is a zero-terminated string, or else memory might get corrupted.

Reply:
So invariants as in function contracts, are one form of validation. But there is also invariant blocks which can also perform this function. Both of which get compiled out in ``-release`` mode.

The reason you would give it ``@system`` is because once initialized you don't want it to become an unknown state. Which would be inherently unsafe. This is what concerns me, *becoming* unsafe, rather than being initialized as unsafe.

It would be the responsibility of the developer to write a constructor should they need more guarantees. I would highly doubt that you will find struct's that have member fields annotated with ``@system`` in the wild today. Meaning if any has been annotated, that it is on purpose.

Regarding Cstring, your example has already solved this and I do not believe you can solve this in the general case. The compiler knows that the pointer does point to something. What it doesn't understand is the requirement for it being null terminated. If you wish to force this (without writing an ``@system`` constructor), you would need to disable the constructor.

Reply:
`invariant` blocks and contracts are completely unrelated, with 'invariants' I mean a compile time guarantee that certain type + bit-pattern combinations never reach `@safe` code, e.g. a bool that is not 0 or 1. While the compiler knows that `bool(3)` is invalid, `@system` fields allow you to make your own types with forbidden bit-patterns, but the compiler can't know what the restrictions are. Therefor it has to be conservative and assume the result of a default constructor is not valid.

Reply:
> This is what concerns me, becoming unsafe, rather than being initialized as unsafe.

Can you give an example of a type that can't be initialized to an invalid value with a default constructor but can be modified to one?

Reply:
Your C string example is a good one.

It has a separate constructor in the form of a free-function that performs the validation.

If you need more guarantees than this provides, you end up having to ``@disable`` the constructor otherwise unsafe code can freely call it and set it into an invalid state.

No matter which way you go, if you want these guarantees as the author of the code, you will have to either disable it or write your own constructor.

Reply:
The C string is not a good example since it _can_ be initialized with an invalid value, which is why the free function needs to be `@trusted`.
It is a case in favor of a `@system` default constructor by default instead of a `@trusted` default constructor.